### DISCUSSION TITLE
where is mixedProductIlJk in RankTwoTensorImplementation.h

### DISCUSSION THREAD

**Discussion by howietu0601** [idaholab/moose#21467] [Q&A General] [answered]
Hi, I just upgraded my moose from the version half years ago, but realize lots of changes happened and my own development is totally messed up now. I used "mixedProductIlJk" in my code which was in RankTwoTensorImplementation.h before but not now; how can I solve this issue? Thanks!

**Comment by ajacquey** [comment, marked as answer]
Hi,

I had the same experience a few days ago.
It seems that some functions (such as `mixedProductXxXx`) have been replaced by a more general approach using `usingTensorIndices` from `MooseTypes` (see [framework/include/utils/MooseTypes.h](https://github.com/idaholab/moose/blob/f1d088405044b932c2090522540fc996d403618a/framework/include/utils/MooseTypes.h))
First, you define the indices:
```
usingTensorIndices(i_, j_, k_, l_);
```

And then the equivalent (someone please correct me if I'm wrong) for two `RankTwoTensor` `a` and `b`:
`a.mixedProductIkJl(b)` becomes `a.times<i_, k_, j_, l_>(b)`
`a.mixedProductJkIl(b)` becomes `a.times<j_, k_, i_, l_>(b)` 
`a.mixedProductIlJk(b)` becomes `a.times<i_, l_, j_, k_>(b)`
and so on...

Hope that helps.
